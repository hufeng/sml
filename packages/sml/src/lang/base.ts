import path from 'node:path'
import { exec } from 'node:child_process'
import Builder from '../common/builder'
import {
  ActorStyleType,
  BaseAst,
  DirectionType,
  GlobalConfigType,
  PackageStyle,
} from './types'

export const noop = () => {}

// ~~~~~~~~~~ builder ~~~~~~~~~~~~~~~~~~~~~~
class ConfigBuilder {
  private config: GlobalConfigType

  constructor(config: GlobalConfigType) {
    this.config = config
  }

  /**
   * setting actor style
   * @param style
   * @returns
   */
  actorStyle(style: ActorStyleType = 'default') {
    this.config.actorStyle = style
    return this
  }

  /**
   * setting package style
   * @param style
   * @returns
   */
  packageStyle(style: PackageStyle = 'Rectangle') {
    this.config.packageStyle = style
    return this
  }

  /**
   * setting direction
   * @param direction
   */
  direction(direction: DirectionType) {
    this.config.direction = direction
  }

  theme(theme: string = 'cerulean-outline') {
    this.config.theme = theme
  }
}

// ~~~~~~~~~~ lang base ~~~~~~~~~~~~~~~~

/**
 * all lang base class
 */
export class Lang<T extends { title: string; config?: GlobalConfigType }> {
  protected meta: T

  constructor(meta: T) {
    this.meta = meta
    this.meta.config = {
      actorStyle: 'default',
      direction: 'left->right',
      packageStyle: 'Rectangle',
      theme: 'cerulean-outline',
    }
  }

  get configuration() {
    return new ConfigBuilder(this.meta.config!)
  }
}

/**
 * all Emitter base class
 */
export abstract class Emitter<T extends BaseAst> {
  protected meta: T
  protected s: Builder

  constructor(meta: T) {
    this.meta = meta
    this.s = new Builder()
  }

  abstract emitCode(): string

  protected buildConfig = (s: Builder) => {
    const { config } = this.meta
    s
      // setting theme
      .str(`!theme ${config!.theme}`)
      // setting actor style
      .ifStr(
        config!.actorStyle !== 'default',
        `skinparam actorStyle ${config!.actorStyle}`,
      )
      .str(`skinparam packageStyle ${config?.packageStyle}`)
      // setting direction
      .str(`${config!.direction.replace('->', ' to ')} direction\n`)
    return s
  }

  protected buildTheme = (s: Builder) => {
    const { config } = this.meta
    s
      // setting theme
      .str(`!theme ${config!.theme}`)
  }

  plantUML(img: string) {
    const jar = path.join(__dirname, '../../bin/plantuml-1.2022.7.jar')

    exec(
      `${this.emitCode()} | java -jar ${jar} -pipe > ${img}`,
      (err, stdout, stderr) => {
        if (err) {
          throw err
        }
        console.log(stdout)
        console.log(stderr)
      },
    )
  }
}
