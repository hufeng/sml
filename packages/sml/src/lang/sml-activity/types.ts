import { BaseAst, LinkContaniner } from '../types'

export type ComponentType =
  | 'participant'
  | 'actor'
  | 'boundary'
  | 'control'
  | 'entity'
  | 'database'
  | 'collections'
  | 'queue'

export interface Component {
  id: string
  label: string
  type: string
}

export type LinkContainer = Omit<Required<LinkContaniner>, 'rels'>

export interface SmlActivityLangMeta extends BaseAst, LinkContainer {
  components: Array<Component>
}

export type ComponentBuilderMeta = SmlActivityLangMeta
