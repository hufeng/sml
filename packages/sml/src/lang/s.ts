export default class S {
  private buff: Array<string | S>

  constructor() {
    this.buff = []
  }

  $s(s: string) {
    this.buff.push(s)
    return this
  }

  $fn(fn: (s: S) => void) {
    fn(this)
    return this
  }

  $if(cond: boolean, str: string) {
    if (cond) {
      this.buff.push(str)
    }
    return this
  }

  $for<T>(arr: Array<T> = [], fn: (s: S, v: T, i?: number) => void) {
    if (arr.length === 0) {
      return this
    }

    arr.forEach((v, i) => {
      fn(this, v, i)
    })
    return this
  }

  $fors<T>(arr: Array<T> = [], fn: (s: S, v: T, i?: number) => void) {
    if (arr.length === 0) {
      return this
    }

    this.buff.push('')
    // append for
    arr.forEach((v, i) => {
      fn(this, v, i)
    })
    this.buff.push('')

    return this
  }

  reset() {
    this.buff = []
    return this
  }

  toString(delimiter = '\n') {
    return this.buff.join(delimiter)
  }
}
