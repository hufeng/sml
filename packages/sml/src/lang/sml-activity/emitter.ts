import S from '../s'
import { Emitter } from '../base'
import { Component, SmlActivityLangMeta } from './types'

export class PumlActivityEmitter extends Emitter<SmlActivityLangMeta> {
  emitCode() {
    const { components, links, vlinks, notes } = this.meta

    return this.s
      .$reset()
      .$s('@startuml')
      .$fn(this.buildTheme)
      .$fors(components, this.buildComponents)
      .$fors(links, this.buildLinks('->'))
      .$fors(vlinks, this.buildVlink('-->'))
      .$fors(notes, this.buildNotes)
      .$s('@enduml')
      .toString()
  }

  buildComponents(s: S, c: Component) {
    s.$s(`${c.type} "${c.label}" as ${c.id}`)
  }
}
