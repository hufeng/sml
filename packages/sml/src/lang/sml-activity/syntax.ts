import { Lang } from '../base'
import { ComponentBuilder } from './builder/component'
import { SmlActivityLangMeta } from './types'

export class SmlActivityLang extends Lang<SmlActivityLangMeta> {
  constructor(meta: SmlActivityLangMeta) {
    super(meta)
  }

  participant(label: string) {
    return new ComponentBuilder(this.meta, label, 'participant')
  }

  actor(label: string) {
    return new ComponentBuilder(this.meta, label, 'actor')
  }

  boundary(label: string) {
    return new ComponentBuilder(this.meta, label, 'boundary')
  }

  control(label: string) {
    return new ComponentBuilder(this.meta, label, 'control')
  }

  entity(label: string) {
    return new ComponentBuilder(this.meta, label, 'entity')
  }

  database(label: string) {
    return new ComponentBuilder(this.meta, label, 'database')
  }

  collections(label: string) {
    return new ComponentBuilder(this.meta, label, 'collections')
  }

  queue(label: string) {
    return new ComponentBuilder(this.meta, label, 'queue')
  }
}
