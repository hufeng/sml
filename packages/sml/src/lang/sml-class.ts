import { Lang } from './sml'
import { globalCollections } from './sml-global'

// ~~~~~~~~~~~~~~ basic type ~~~~~~~~~~~~~~~~~
type DataType = string
export type VisibleType = 'private' | 'protected' | 'public' | 'package_private'
type ParamType = { name: string; type: DataType }
type MethodOptional = (m: MethodType) => void
type VisibleOptional = (v: { visible: VisibleType }) => void
type abstractOptional = (a: { abstract: boolean }) => void

// ~~~~~~~~~~~~~~ container type ~~~~~~~~~~~~
type FiledType = {
  name: string
  type: DataType
  visible: VisibleType
}
type MethodType = {
  abstract: boolean
  visible: VisibleType
  name: string
  params: Array<ParamType>
  ret: DataType
}
type ClazzType = {
  name: string
  abstract: boolean
  fields: Array<FiledType>
  methods: Array<MethodType>
  extends: Array<string>
  implements: Array<string>
}
type InterfaceType = {
  name: string
  implements: Array<string>
  methods: Array<MethodType>
}
type EnumFieldTyep = {
  name: string
  value?: number | string
}
type EnumType = {
  name: string
  fields: Array<EnumFieldTyep>
}
type StructType = ClazzType
type ProtocolType = InterfaceType
export interface SmlClazzMeta {
  clazzes: Array<ClazzType>
  interfaces: Array<InterfaceType>
  enums: Array<EnumType>
  structs: Array<StructType>
  protocols: Array<ProtocolType>
}

//~~~~~~~~~~~~ builder ~~~~~~~~~~~~~~~~
class ClazzBuilder {
  constructor(private clazz: ClazzType) {}

  field(name: string, type: DataType, visible: VisibleType = 'private') {
    this.clazz.fields.push({ name, type, visible })
    return this
  }

  method(
    name: string,
    params: MethodOptional,
    ret: MethodOptional,
    ...rest: Array<VisibleOptional | abstractOptional>
  ) {
    const method = {
      abstract: false,
      visible: 'public' as VisibleType,
      name,
      params: [] as Array<ParamType>,
      ret: '' as DataType,
    }
    params(method)
    ret(method)
    rest.forEach((opt) => opt(method))
    this.clazz.methods.push(method)
    return this
  }

  extends(...names: Array<string>) {
    this.clazz.extends = [...this.clazz.extends, ...names]
    return this
  }

  implements(...names: Array<string>) {
    this.clazz.implements = [...this.clazz.implements, ...names]
    return this
  }
}

class InfBuilder {
  constructor(private interfaces: InterfaceType) {}

  method(
    name: string,
    args: MethodOptional = (m: MethodType) => (m.params = []),
    ret: MethodOptional = (m: MethodType) => (m.ret = 'void'),
  ) {
    const method = {
      name,
      visible: 'public' as VisibleType,
      abstract: false,
      params: [] as Array<ParamType>,
      ret: '' as DataType,
    }
    args(method)
    ret(method)
    this.interfaces.methods.push(method)
    return this
  }

  implements(...names: Array<string>) {
    this.interfaces.implements = [...this.interfaces.implements, ...names]
    return this
  }
}

class EnumBuilder {
  constructor(private e: EnumType) {}

  field(name: string, value?: string | number) {
    this.e.fields.push({ name, value })
    return this
  }
}

// ~~~~~~~~~~~~~ define class lang modeing ~~~~~~~~~~~~~~~~~~~
enum T {
  void = 'void',
  string = `string`,
  boolean = `boolean`,
  number = `number`,

  go_int8 = `go_int8`,
  go_int16 = `go_int16`,
  go_int32 = `go_int32`,
  go_int = `go_int`,
  go_uint = `go_uint`,
  go_byte = 'go_byte',
  go_int64 = `go_int64`,
  go_uint8 = `go_uint8`,
  go_uint16 = `go_uint16`,
  go_uint32 = `go_uint32`,
  go_uint64 = `go_uint64`,
  go_float32 = `go_float32`,
  go_float64 = `go_float64`,

  py_int = 'py_int',
  py_bool = `py_bool`,
  py_float = `py_float`,

  java_byte = `java_byte`,
  java_short = `java_short`,
  java_int = `java_int`,
  java_long = `java_long`,
  java_float = `java_float`,
  java_double = `java_double`,
  java_boolean = `java_boolean`,
  java_char = `java_char`,
}
export class SmlClazzLang extends Lang {
  private meta: SmlClazzMeta

  constructor(title: string) {
    super(title)
    this.meta = {
      clazzes: [],
      interfaces: [],
      enums: [],
      structs: [],
      protocols: [],
    }
  }

  /**
   * define abstract class
   * @param name abstract class name
   * @returns
   */
  abstractClazz(name: string) {
    const clazz = {
      name,
      abstract: true,
      fields: [],
      methods: [],
      extends: [],
      implements: [],
    }
    this.meta.clazzes.push(clazz)
    return new ClazzBuilder(clazz)
  }

  /**
   * define class
   * @param name class name
   * @returns
   */
  clazz(name: string) {
    const clazz = {
      name,
      abstract: false,
      fields: [],
      methods: [],
      extends: [],
      implements: [],
    }
    this.meta.clazzes.push(clazz)
    return new ClazzBuilder(clazz)
  }

  /**
   * define interface
   * @param name interface name
   * @returns
   */
  interface(name: string) {
    const inf = {
      name,
      implements: [],
      methods: [],
    } as InterfaceType
    this.meta.interfaces.push(inf)
    return new InfBuilder(inf)
  }

  /**
   * define enum
   * @param name enum name
   * @returns
   */
  enum(name: string) {
    const e = {
      name,
      fields: [],
    }
    this.meta.enums.push(e)
    return new EnumBuilder(e)
  }

  /**
   * define struct
   * @param name struct name
   * @returns
   */
  struct(name: string) {
    const s = {
      name,
      abstract: false,
      fields: [],
      methods: [],
      extends: [],
      implements: [],
    } as ClazzType
    this.meta.structs.push(s)
    return new ClazzBuilder(s)
  }

  /**
   * define a protocol
   * @param name protocol name
   * @returns
   */
  protocol(name: string) {
    const prot = {
      name,
      implements: [],
      methods: [],
    } as InterfaceType
    this.meta.protocols.push(prot)
    return new InfBuilder(prot)
  }

  /**
   * args optional
   * @param args
   * @returns
   */
  args(...args: Array<ParamType>) {
    return (m: MethodType) => (m.params = [...m.params, ...args])
  }

  /**
   * arg optional
   */
  arg(name: string, type: string) {
    return { name, type } as ParamType
  }

  /**
   * return value type
   * @param type
   * @returns
   */
  ret(type: string) {
    return (m: MethodType) => (m.ret = type)
  }

  /**
   * setting field or method visible
   * @param v
   * @returns
   */
  visible(v: VisibleType) {
    return (val: { visible: VisibleType }) => (val.visible = v)
  }

  /**
   * setting method whether is abstract
   * @param v
   * @returns
   */
  abstract(v: boolean) {
    return (val: { abstract: boolean }) => (val.abstract = v)
  }

  t = T
}

/**
 * ClassDiagram factory
 * @param title
 * @param fn
 * @returns
 */
export function ClassDiagram(title: string, fn: (ml: SmlClazzLang) => void) {
  const lang = new SmlClazzLang(title)
  fn(lang)
  globalCollections.add(lang)
  return lang
}
