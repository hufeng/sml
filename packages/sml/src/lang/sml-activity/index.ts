import { PumlActivityEmitter } from './emitter'
import { SmlActivityLang } from './syntax'

export function SmlActivityDiagram(
  title: string,
  fn: (ml: SmlActivityLang) => void,
) {
  const ast = {
    title,
    components: [],
    links: [],
    vlinks: [],
    notes: [],
  }

  fn(new SmlActivityLang(ast))

  return { ast, emitter: new PumlActivityEmitter(ast) }
}
