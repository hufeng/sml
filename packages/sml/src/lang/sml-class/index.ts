import { PumlClassEmitter } from './emitter'
import { SmlClazzLang as SmlClassLang } from './syntax'

/**
 * ClassDiagram factory
 * @param title
 * @param fn
 * @returns
 */
export function ClassDiagram(title: string, fn: (ml: SmlClassLang) => void) {
  const ast = {
    title,
    clazzes: [],
    interfaces: [],
    enums: [],
    structs: [],
    protocols: [],
  }

  fn(new SmlClassLang(ast))

  return {
    ast,
    emitter: new PumlClassEmitter(ast),
  }
}
