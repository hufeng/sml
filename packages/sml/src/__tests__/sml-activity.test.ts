import { SmlActivityDiagram } from '../lang/sml-activity'

describe('sml activity test suite', () => {
  it('test syntax', () => {
    const { ast, emitter } = SmlActivityDiagram('hello activities', (ml) => {
      const alice = ml.actor('Alice')
      const b = ml.boundary('Boundary')
      const c = ml.control('Control')
      const e = ml.entity('Entity')
      const db = ml.database('Database')
      const coll = ml.collections('Collections')
      const queue = ml.queue('Queue')

      alice.link(b, (l) => l.commentOf('To Boundary'))
      b.link(c)
      c.link(e)
      e.link(db)
      db.link(coll)
      coll.link(queue)
      queue.vlink(coll)
    })

    expect(ast).toMatchInlineSnapshot(`
      {
        "components": [
          {
            "id": "a64489c85",
            "label": "Alice",
            "type": "actor",
          },
          {
            "id": "be234de58",
            "label": "Boundary",
            "type": "boundary",
          },
          {
            "id": "ca1595abb",
            "label": "Control",
            "type": "control",
          },
          {
            "id": "e1a434bef",
            "label": "Entity",
            "type": "entity",
          },
          {
            "id": "de307db07",
            "label": "Database",
            "type": "database",
          },
          {
            "id": "ca9fc9193",
            "label": "Collections",
            "type": "collections",
          },
          {
            "id": "q722ad2d0",
            "label": "Queue",
            "type": "queue",
          },
        ],
        "config": {
          "actorStyle": "awesome",
          "direction": "left->right",
          "packageStyle": "Rectangle",
          "theme": "sketchy-outline",
        },
        "links": [
          {
            "comment": "To Boundary",
            "from": "a64489c85",
            "to": [
              "be234de58",
            ],
          },
          {
            "from": "be234de58",
            "to": [
              "ca1595abb",
            ],
          },
          {
            "from": "ca1595abb",
            "to": [
              "e1a434bef",
            ],
          },
          {
            "from": "e1a434bef",
            "to": [
              "de307db07",
            ],
          },
          {
            "from": "de307db07",
            "to": [
              "ca9fc9193",
            ],
          },
          {
            "from": "ca9fc9193",
            "to": [
              "q722ad2d0",
            ],
          },
        ],
        "notes": [],
        "rels": [],
        "title": "hello activities",
        "vlinks": [
          {
            "from": "q722ad2d0",
            "to": [
              "ca9fc9193",
            ],
          },
        ],
      }
    `)
    expect(emitter.emitCode()).toMatchSnapshot()
  })
})
