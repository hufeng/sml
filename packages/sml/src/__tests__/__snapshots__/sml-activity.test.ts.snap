// Vitest Snapshot v1

exports[`sml activity test suite > test syntax 2`] = `
"@startuml
!theme sketchy-outline

actor \\"Alice\\" as a64489c85
boundary \\"Boundary\\" as be234de58
control \\"Control\\" as ca1595abb
entity \\"Entity\\" as e1a434bef
database \\"Database\\" as de307db07
collections \\"Collections\\" as ca9fc9193
queue \\"Queue\\" as q722ad2d0


a64489c85 -> be234de58 : To Boundary
be234de58 -> ca1595abb
ca1595abb -> e1a434bef
e1a434bef -> de307db07
de307db07 -> ca9fc9193
ca9fc9193 -> q722ad2d0


q722ad2d0 --> ca9fc9193

@enduml"
`;
